### PR TITLE
fix bug: wrong output dimension when "keep_dims" is false in pooling layer.

### PR DESCRIPTION
**Pull request by Crayon-new**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/932

### Pull Request Readiness Checklist

Fixed bug in #20896：When parsing a pooling layer in dnn/tensorflow. If keep_dims is false, the  additional "nhwc" layer and "squeeze" layer will be lost. Because the final "squeeze" layer name is not the original layer name. 

Solution: 
Change the final output layer's name to the orignal **pooling** layer's name.

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
opencv_extra=tf_pooling
```

### CODE REVIEW

**Comment by Crayon-new** [comment]
@asmorkalov Hello, anything else do I need to do?

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 1]
```diff
@@ -2371,24 +2371,34 @@ void TFImporter::parseMean(tensorflow::GraphDef& net, const tensorflow::NodeDef&
...
+                int id = dstNet.addLayer(name, "Pooling", layerParams);
+                layer_id[name] = id;
+                connect(layer_id, dstNet, parsePin(layer.input(0)), id, 0);
+            }
+            else
             {
                 // To keep correct order after squeeze dims we first need to change layout from NCHW to NHWC
+                layerParams.set("pool", pool_type);
+                layerParams.set(axis == 2 ? "kernel_w" : "kernel_h", 1);
+                layerParams.set(axis == 2 ? "global_pooling_h" : "global_pooling_w", true);
```
Comment (rogday) [inline modules/dnn/src/tensorflow/tf_importer.cpp:2387]: Seems like this bit of code can be extracted from the if body.

**Review by rogday — COMMENTED** [review]
Thank you for your contribution! IIRC, bugfixes should go to 3.4 branch and not 4.x, so, please, rebase.

I think, there is more to it. Could you check other cases as well? I used `+1` instead of `ExpandDims` to avoid more issues with dimensions.

I used the following code for tests, it shows errors with axis=[0], [3](keepdims true and false), [1,2]:

```python

axises = [[0], [1], [2], [3], [1, 2]]

for axis in axises:
    for keepdims in [False, True]:
        inp = tf.placeholder(tf.float32, [2, 3, 4, 1])
        biasadd = tf.nn.bias_add(inp, [1], data_format='NHWC')
        print(axis, keepdims)
        reduced = tf.reduce_sum(biasadd, axis=axis, keepdims=keepdims)
        save(inp, reduced + 1, f'reduce_sum_{axis}_{keepdims}')
 
 ```

```cpp

TEST_P(Test_TensorFlow_layers, pooling_reduce_sum3)
{
    std::vector<std::vector<int>> axises = {{0}, {1}, {2}, {3}, {1, 2}};

    for (const auto& axis : axises)
    {
        for (int keepdims = 0; keepdims <= 1; ++keepdims)
        {
            std::stringstream ss;
            ss << "reduce_sum_[" << axis[0];
            if (axis.size() > 1)
            {
                ss << ", " << axis[1];
            }
            ss << "]_" << (keepdims ? "True" : "False");
            std::cout << ss.str() << std::endl;
            try
            {
                runTensorFlowNet(ss.str());
            }
            catch (const std::exception& e)
            {
                std::cout << e.what() << std::endl;
            }
        }
    }
}

```

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 2]
```diff
@@ -2288,18 +2310,26 @@ void TFImporter::parseMean(tensorflow::GraphDef& net, const tensorflow::NodeDef&
...
+        else
+        {
+            std::string poolingName = name+"/Pooling";
+            int id = dstNet.addLayer(poolingName, "Pooling", layerParams);
+            layer_id[poolingName] = id;
+            connect(layer_id, dstNet, parsePin(layer.input(0)), id, 0);
             LayerParams flattenLp;
-            std::string flattenName = name + "/flatten";
+            std::string flattenName = name;
             CV_Assert(layer_id.find(flattenName) == layer_id.end());
```
Comment (rogday) [inline modules/dnn/src/tensorflow/tf_importer.cpp:2328]: Consider using `const std::string&` and removing the assert.

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 3]
```diff
@@ -2254,32 +2274,34 @@ void TFImporter::parseMean(tensorflow::GraphDef& net, const tensorflow::NodeDef&
...
+            int id = dstNet.addLayer(poolingName, "Pooling", layerParams);
+            layer_id[poolingName] = id;
+            connect(layer_id, dstNet, Pin(permName), id, 0);
 
             if (!keepDims)
             {
                 LayerParams squeezeLp;
-                std::string squeezeName = name + "/squeeze";
+                std::string squeezeName = name;
                 CV_Assert(layer_id.find(squeezeName) == layer_id.end());
```
Comment (rogday) [inline modules/dnn/src/tensorflow/tf_importer.cpp:2292]: Consider changing the type to `const std::string&` and removing the assert.

File: modules/dnn/test/test_tf_importer.cpp [hunk 4]
```diff
@@ -413,6 +413,50 @@ TEST_P(Test_TensorFlow_layers, pooling_reduce_sum)
...
+            {
+                runTensorFlowNet(ss.str());
+            }
+            catch (const std::exception& e)
+            {
+                std::cout << e.what() << std::endl;
+            }
+        }
+    }
+}
```
Comment (rogday) [inline modules/dnn/test/test_tf_importer.cpp:458]: My proposed solution was a draft, which helps debugging the problem.
Consider this instead:

```cpp

TEST_P(Test_TensorFlow_layers, pooling_reduce_sum2)
{
    int axises[] = {0, 1, 2, 3};
    for (int keepdims = 0; keepdims <= 1; ++keepdims)
    {
        for (int i = 0; i < sizeof(axises)/sizeof(axises[0]); ++i)
        {
            runTensorFlowNet(cv::format("reduce_sum_[%d]_%s", axises[i], (keepdims ? "True" : "False")));
        }
        runTensorFlowNet(cv::format("reduce_sum_[1, 2]_%s", keepdims ? "True" : "False"));
    }
}

```

**Review by rogday — CHANGES_REQUESTED** [review]
Good work! :+1:

File: modules/dnn/test/test_tf_importer.cpp [hunk 5]
```diff
@@ -413,6 +413,19 @@ TEST_P(Test_TensorFlow_layers, pooling_reduce_sum)
...
+TEST_P(Test_TensorFlow_layers, pooling_reduce_sum2)
+{
+    int axises[] = {0, 1, 2, 3};
+    for (int keepdims = 0; keepdims <= 1; ++keepdims)
+    {
+        for (int i = 0; i < sizeof(axises)/sizeof(axises[0]); ++i)
+        {
+            runTensorFlowNet(cv::format("reduce_sum_[%d]_%s", axises[i], (keepdims ? "True" : "False")));
+        }
+        runTensorFlowNet(cv::format("reduce_sum_[1, 2]_%s", keepdims ? "True" : "False"));
```
Comment (alalek) [inline modules/dnn/test/test_tf_importer.cpp:425]: > `'['`
> `']'`
> `','`
> `' '`

Could you please use sanitized file names? (please use `_` instead of these symbols or space)

File: modules/dnn/test/test_tf_importer.cpp (hunk 4 above)
Comment (Crayon-new) [reply modules/dnn/test/test_tf_importer.cpp]: fixed

File: modules/dnn/src/tensorflow/tf_importer.cpp (hunk 3 above)
Comment (Crayon-new) [reply modules/dnn/src/tensorflow/tf_importer.cpp]: fixed

File: modules/dnn/src/tensorflow/tf_importer.cpp (hunk 2 above)
Comment (Crayon-new) [reply modules/dnn/src/tensorflow/tf_importer.cpp]: fixed

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 6]
```diff
@@ -2288,18 +2310,26 @@ void TFImporter::parseMean(tensorflow::GraphDef& net, const tensorflow::NodeDef&
...
+        
+        if (keepDims)
         {
+            int id = dstNet.addLayer(name, "Pooling", layerParams);
+            layer_id[name] = id;
+            connect(layer_id, dstNet, parsePin(layer.input(0)), id, 0);
+        }
+        else
+        {
+            std::string poolingName = name+"/Pooling";
```
[the remark this answers is not in the document]
Comment (Crayon-new) [reply modules/dnn/src/tensorflow/tf_importer.cpp]: fixed

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 7]
```diff
@@ -2227,21 +2238,30 @@ void TFImporter::parseMean(tensorflow::GraphDef& net, const tensorflow::NodeDef&
...
-                std::string permName = name + "/nchw";
-                Pin inpId = Pin(name);
+                std::string permName = name + "/nhwc";
+                Pin inpId = Pin(poolingName);
                 addPermuteLayer(order, permName, inpId);
 
                 LayerParams squeezeLp;
-                std::string squeezeName = name + "/squeeze";
+                std::string squeezeName = name;
                 CV_Assert(layer_id.find(squeezeName) == layer_id.end());
```
[the remark this answers is not in the document]
Comment (Crayon-new) [reply modules/dnn/src/tensorflow/tf_importer.cpp]: fixed

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 8]
```diff
@@ -2254,32 +2274,34 @@ void TFImporter::parseMean(tensorflow::GraphDef& net, const tensorflow::NodeDef&
...
+            std::string permName = name + "/nhwc";
+            addPermuteLayer(order, permName, inpId);
 
             layerParams.set("pool", pool_type);
             layerParams.set("kernel_h", 1);
             layerParams.set("global_pooling_w", true);
-            int id = dstNet.addLayer(name, "Pooling", layerParams);
-            layer_id[name] = id;
-            connect(layer_id, dstNet, inpId, id, 0);
+            std::string poolingName = name+"/Pooling";
```
[the remark this answers is not in the document]
Comment (Crayon-new) [reply modules/dnn/src/tensorflow/tf_importer.cpp]: fixed

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 9]
```diff
@@ -2227,21 +2238,30 @@ void TFImporter::parseMean(tensorflow::GraphDef& net, const tensorflow::NodeDef&
...
                 int order[] = {0, 2, 3, 1};  // From OpenCV's NCHW to NHWC.
-                std::string permName = name + "/nchw";
-                Pin inpId = Pin(name);
+                std::string permName = name + "/nhwc";
+                Pin inpId = Pin(poolingName);
                 addPermuteLayer(order, permName, inpId);
 
                 LayerParams squeezeLp;
-                std::string squeezeName = name + "/squeeze";
+                std::string squeezeName = name;
```
[the remark this answers is not in the document]
Comment (Crayon-new) [reply modules/dnn/src/tensorflow/tf_importer.cpp]: fixed

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 10]
```diff
@@ -2227,21 +2238,30 @@ void TFImporter::parseMean(tensorflow::GraphDef& net, const tensorflow::NodeDef&
...
+            if (keepDims)
+            {
+                int id = dstNet.addLayer(name, "Pooling", layerParams);
+                layer_id[name] = id;
+                connect(layer_id, dstNet, parsePin(layer.input(0)), id, 0);
+            }
+            else
             {
                 // To keep correct order after squeeze dims we first need to change layout from NCHW to NHWC
+                std::string poolingName = name+"/Pooling";
```
[the remark this answers is not in the document]
Comment (Crayon-new) [reply modules/dnn/src/tensorflow/tf_importer.cpp]: fixed

**Review by rogday — APPROVED** [review]
LGTM :+1:
